File: day11/part2/main.dart
// -*- compile-command: "pub get && dart main.dart ../example1.txt ../input.txt"; -*-

import 'dart:io';

import 'package:collection/equality.dart';
import 'package:tuple/tuple.dart';

main(List<String> args) {
  for (var arg in args) {
    process(arg);
  }

  print('Done.');
}

process(String filename) {
  print('Processing $filename ...');

  var puz = Puzzle(filename);

  while (true) {
    var newGrid = puz.iterate();
    if (MapEquality().equals(puz.grid, newGrid)) {
      break;
    }
    puz.grid = newGrid;
  }

  var occupied = puz.occupied();

  print('Solution: $occupied');
}

class Puzzle {
  int width = 0;
  int height = 0;
  var grid = Map<Tuple2<int,int>,String>();

  Puzzle(String filename) {
    var lines = File(filename).readAsLinesSync();
    for (var line in lines) {
      this.width = line.length;
      var chars = line.split('');
      for (var x = 0; x < chars.length; x++) {
        if (chars[x] == 'L') {
          var key = Tuple2<int,int>(x, this.height);
          this.grid[key] = 'L';
        }
      }
      this.height++;
    }
  }

  Map<Tuple2<int,int>,String> iterate() {
    var r = Map<Tuple2<int,int>,String>();
    grid.forEach((k, v) {
        var adj = countAdjacentOccupied(k);
        if (v == 'L' && adj == 0) {
          r[k] = '#';
        } else if (v == '#' && adj >= 5) {
          r[k] = 'L';
        } else {
          r[k] = v;
        }
    });
    return r;
  }

  bool adjacent(int x, int dx, int y, int dy) {
    var u = x + dx;
    var v = y + dy;
    var adj = 0;
    while (u >= 0 && v >= 0 && u < width && v < height) {
      var key = Tuple2<int,int>(u, v);
      var val = grid[key];
      if (val == '#') {
        return true;
      }
      if (val == 'L') {
        return false;
      }
      u += dx;
      v += dy;
    }
    return false;
  }

  int countAdjacentOccupied(Tuple2<int,int> k) {
    var x = k.item1;
    var y = k.item2;
    var adj = 0;

    if (adjacent(x,-1, y,-1)) adj++;
    if (adjacent(x, 0, y,-1)) adj++;
    if (adjacent(x, 1, y,-1)) adj++;

    if (adjacent(x,-1, y, 0)) adj++;
    if (adjacent(x, 1, y, 0)) adj++;

    if (adjacent(x,-1, y, 1)) adj++;
    if (adjacent(x, 0, y, 1)) adj++;
    if (adjacent(x, 1, y, 1)) adj++;

    return adj;
  }

  int occupied() {
    var count = 0;
    grid.forEach((k, v) {
        if (v == '#') count++;
    });
    return count;
  }
}

File: day11/part2/main.go
// -*- compile-command: "go run main.go ../example1.txt ../input.txt"; -*-

package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"strings"
)

var (
	verbose = flag.Bool("v", false, "Verbose log messages")
)

func main() {
	flag.Parse()

	for _, arg := range flag.Args() {
		process(arg)
	}

	log.Printf("Done.")
}

func process(filename string) {
	puz := readPuzzle(filename)

	for puz.iterate() {
	}

	occupied := puz.occupied()

	log.Printf("Solution: %v", occupied)
}

func (p *Puzzle) occupied() int {
	var count int
	for _, v := range p.grid {
		if v == "#" {
			count++
		}
	}
	return count
}

func (p *Puzzle) iterate() bool {
	r := map[string]string{}
	var changesMade bool
	for k, v := range p.grid {
		adj := p.countAdjacentOccupied(k)
		// log.Printf("k=%v, v=%v, adj=%v", k, v, adj)
		if v == "L" && adj == 0 {
			r[k] = "#"
			changesMade = true
		} else if v == "#" && adj >= 5 {
			r[k] = "L"
			changesMade = true
		} else {
			r[k] = v
		}
	}

	if changesMade {
		p.grid = r
	}

	return changesMade
}

func (p *Puzzle) countAdjacentOccupied(k string) int {
	x, y := p.parseKey(k)
	var adj int

	t := func(dx, dy int) {
		u, v := x+dx, y+dy
		for u >= 0 && v >= 0 && u < p.width && v < p.height {
			val := p.lookup(u, v)
			if val == "#" {
				adj++
				return
			}
			if val == "L" {
				return
			}
			u += dx
			v += dy
		}
	}

	t(-1, -1)
	t(0, -1)
	t(+1, -1)
	t(-1, 0)

	t(+1, 0)
	t(-1, +1)
	t(0, +1)
	t(+1, +1)

	return adj
}

type Puzzle struct {
	width  int
	height int
	grid   map[string]string
	keyX   map[string]int
	keyY   map[string]int
}

func (p *Puzzle) String() string {
	var s string
	for y := 0; y < p.height; y++ {
		for x := 0; x < p.width; x++ {
			v := p.lookup(x, y)
			if v == "" {
				s += "."
			} else {
				s += v
			}
		}
		s += "\n"
	}
	return s
}

func readPuzzle(filename string) *Puzzle {
	logf("Processing %v ...", filename)
	buf, err := ioutil.ReadFile(filename)
	check("ReadFile: %v", err)

	puz := &Puzzle{grid: map[string]string{}, keyX: map[string]int{}, keyY: map[string]int{}}
	lines := strings.Split(string(buf), "\n")
	for y, line := range lines {
		if line == "" {
			continue
		}
		puz.width = len(line)
		puz.height++
		for x := 0; x < len(line); x++ {
			if line[x:x+1] == "L" {
				key := genKey(x, y)
				puz.grid[key] = "L"
				puz.keyX[key] = x
				puz.keyY[key] = y
			}
		}
	}

	return puz
}

func (p *Puzzle) lookup(x, y int) string {
	return p.grid[genKey(x, y)]
}

func genKey(x, y int) string {
	return fmt.Sprintf("%v,%v", y, x)
}

func (p *Puzzle) parseKey(key string) (x, y int) {
	return p.keyX[key], p.keyY[key]
}

func check(fmtStr string, args ...interface{}) {
	if err := args[len(args)-1]; err != nil {
		log.Fatalf(fmtStr, args...)
	}
}

func logf(fmt string, args ...interface{}) {
	if *verbose {
		log.Printf(fmt, args...)
	}
}
